src/api/stats/common/getCompoundV2Apys.ts

```diff
@@ -0,0 +1,271 @@
+
+  const supplyApys = data.supplyRates.map(v => v.times(BLOCKS_PER_YEAR).div('1e18'));
+  const borrowApys = data.borrowRates.map(v => v.times(BLOCKS_PER_YEAR).div('1e18'));
+
+  const annualCompSupplyInUsd = data.compSupplySpeeds.map(v =>
+    v.times(BLOCKS_PER_YEAR).div(compDecimals).times(compPrice)
+  );
+  const annualCompBorrowInUsd = data.compBorrowSpeeds.map(v =>
+    v.times(BLOCKS_PER_YEAR).div(compDecimals).times(compPrice)
+  );
+
+  const totalSuppliesInUsd = data.totalSupplies.map((v, i) =>
+    v
+      .times(data.exchangeRatesStored[i])
+      .div('1e10')
+      .div(params.pools[i].decimals)
+      .div(data.cTokenDecimals[i])
+      .times(data.tokenPrices[i])
+  );
+  const totalBorrowsInUsd = data.totalBorrows.map((v, i) =>
+    v.times(data.tokenPrices[i]).div(params.pools[i].decimals)
+  );
+
+  const supplyCompApys = annualCompSupplyInUsd.map((v, i) => v.div(totalSuppliesInUsd[i]));
+  const borrowCompApys = annualCompBorrowInUsd.map((v, i) => v.div(totalBorrowsInUsd[i]));
+
+  return {
+    supplyApys,
+    borrowApys,
+    supplyCompApys,
+    borrowCompApys,
+  };
+};
+
+const getPoolLeveragedApy = (
+  pool: CompoundV2Pool,
+  supplyApy: BigNumberStatic,
+  borrowApy: BigNumberStatic,
+  supplyCompApy: BigNumberStatic,
+  borrowCompApy: BigNumberStatic
+) => {
+  const { leveragedSupplyBase, leveragedBorrowBase, leveragedSupplyComp, leveragedBorrowComp } =
+    getLeveragedApys(
+      supplyApy,
+      borrowApy,
+      supplyCompApy,
+      borrowCompApy,
+      pool.borrowDepth,
+      pool.borrowRate
+    );
+  const totalComp = leveragedSupplyComp.plus(leveragedBorrowComp);
+  const shareAfterBeefyPerformanceFee = 1 - getTotalPerformanceFeeForVault(pool.name);
+  const compoundedComp = compound(totalComp, BASE_HPY, 1, shareAfterBeefyPerformanceFee);
+  const apy = leveragedSupplyBase.minus(leveragedBorrowBase).plus(compoundedComp).toNumber();
+
+  return { [pool.name]: apy };
+};
+
+const getPoolsData = async (params: CompoundV2ApyParams): Promise<PoolsData> => {
+  const comptrollerAbi = params.comptrollerAbi ?? IComptroller;
+  const cTokenAbi = params.cTokenAbi ?? IToken;
+
+  const comptrollerContract = getContract(comptrollerAbi, params.comptroller);
+  const multicall = new MultiCall(params.web3 as any, multicallAddress(params.chainId));
+
+  const supplyRateCalls = [];
+  const borrowRateCalls = [];
+  const compSupplySpeedCalls = [];
+  const compBorrowSpeedCalls = [];
+  const totalSupplyCalls = [];
+  const totalBorrowsCalls = [];
+  const exchangeRateStoredCalls = [];
+  const cTokenDecimalsCalls = [];
+
+  let promises = [];
+  params.pools.forEach(pool =>
+    promises.push(fetchPrice({ oracle: pool.oracle, id: pool.oracleId }))
+  );
+  const tokenPrices: BigNumber[] = await Promise.all(promises);
+
+  params.pools.forEach(pool => {
+    const cTokenContract = getContract(cTokenAbi, pool.cToken);
+    supplyRateCalls.push({ supplyRate: cTokenContract.methods.supplyRatePerBlock() });
```

Comment:
https://github.com/beefyfinance/beefy-api/blob/704761bc1ac86357dd5ae7dfc54c1d1406b7ab69/src/api/stats/bsc/getDotDotApy.js#L33-L49

can call multiple methods in 1 call object. doesnt change anything and no need to update but could be handy next time. i think it looks simpler, thats why i suggested supplyCalls and borrowCalls

Reply:
Yes, I think that would probably have been cleaner.  Honestly, I just went with exactly the form that was done in `getMasterChefApys.ts` Another issue I ran into for Sonne specifically is that the `compSpeeds()` function we have used in the past (Scream, etc) returns zero for Sonne.  Maybe the next platform will be different again... Perhaps the dream of a common CompoundV2 implementation is too ambitious.